doc/coding_guide/Coding_guidelines.md

```diff
@@ -0,0 +1,650 @@
+
+    Wrong:
+```
+    mfxStatus sts = MFX_ERR_NONE;
+
+    mfxU32 buffer[100];
+
+    // 50 lines of some code without sts and buffer usage
+
+    sts = WriteDataToBuffer(buffer);
+
+    // 50 more lines
+
+    return sts;
+```
+    Correct:
+```
+    // a lot of code
+
+    mfxU32 buffer[100];
+
+    mfxStatus sts = WriteDataToBuffer(buffer);
+    MFX_CHECK_STS(sts);
+
+    // a lot of code
+```
+-   Prefer switch operator to giant else-if statements.
+
+2.  Keep code at level of 0 compiler warning build.
+
+3.  Do not rely on side effects of your code.
+
+**Code comments**
+
+4.  Do not leave dead commented code. It often will be barriers in source base forever. Either leave meaningful comment why this code is commented (i.e. determine the condition when this code can be switched on again) otherwise delete it at all. Traces of deleted code will remain in git history.
+
+**Includes, namespaces**
+
+5.  Watch your includes.
+
+-   Limit it to bare minimum.
+
+-   Prefer limiting scope of include. Do not put everything to some common header which will propagate it's content to all translation units which don't need it.
+
+-   Try to capture the balance between spreading headers to many source files, which makes dependency graph more complicated and accumulating many of the headers to some common include header, which may lead to significant code size inflation after preprocessing.
+
+-   Prefer to use forward declaration instead of including a header if you don't need complete type.
+
+-   Do not put using declarations in headers. It could lead to undesired scope expansion.
+
+**Macro**
+
+6.  Use macro only if you want some compile-time instrumentation.
+
+-   Other ways use functions, templates and constants.
+
+    Example:
+```
+    // Instead of
+    #define BLOCK_SIZE 1024
+
+    // Use
+    constexpr BLOCK_SIZE = 1024;
+```
+-   Remember when you pass something to macro, it will be used in the way you provided it and will be expanded textually.
+
+    Example:
+```
+    IPP_MAX(++i, j);
+    // will become
+    ((++i) > (j) ? (++i) : j) // twice incremented, it is an unwanted side effect
+```
+-   Use ```MFX_CHECK``` / ```MFX_RETURN``` / ```MFX_CHECK_STS``` macro for ```mfxStatus``` checks. It allows to turn on tracing.
+
+    Hint: Treat these macros like an ```assert``` -  expression have to be evaluated to true and no action is taken in that case, otherwise error is returned.
+
+    Do not put complex conditions into these macros it may reduce readability, if you need such conditions use the following approach:
+```
+    bool mem_pattern_valid = (par->IOPattern & MFX_IOPATTERN_OUT_VIDEO_MEMORY) && (par->IOPattern & MFX_IOPATTERN_OUT_SYSTEM_MEMORY));
+    MFX_CHECK(mem_pattern_valid, MFX_ERR_INVALID_VIDEO_PARAM);
+```
+    Do not put function calls inside these macro, it may lead to double call.
```

Comment:
Something is wring with formatting here (visible when rendered: https://github.com/mgonchar/MediaSDK/blob/introduce_guidelines/doc/coding_guide/Coding_guidelines.md). Probably you need to add empty lines before/after coding blocks.

Reply:
tried to fix all the formatting in next patch